fastplotlib/layouts/_frame/_jupyter_output.py

```diff
@@ -0,0 +1,84 @@
+from typing import *
+
+from ipywidgets import VBox, Widget
+from sidecar import Sidecar
+from IPython.display import display
+
+from ._ipywidget_toolbar import IpywidgetToolBar
+
+
+class JupyterOutputContext(VBox):
+    """
+    Output context to display plots in jupyter. Inherits from ipywidgets.VBox
+
+    Basically vstacks plot canvas, toolbar, and other widgets. Uses sidecar if desired.
+    """
+    def __init__(
+            self,
+            frame,
+            make_toolbar: bool,
+            use_sidecar: bool,
+            sidecar_kwargs: dict,
+            add_widgets: List[Widget],
+    ):
+        """
+
+        Parameters
+        ----------
+        frame:
+            Plot frame for which to generate the output context
+
+        sidecar_kwargs: dict
+            optional kwargs passed to Sidecar
+
+        add_widgets: List[Widget]
+            list of ipywidgets to stack below the plot and toolbar
+        """
+        self.frame = frame
+        self.toolbar = None
+        self.sidecar = None
+
+        # verify they are all valid ipywidgets
+        if False in [isinstance(w, Widget) for w in add_widgets]:
+            raise TypeError(
+                f"add_widgets must be list of ipywidgets, you have passed:\n{add_widgets}"
+            )
+
+        self.use_sidecar = use_sidecar
+
+        if not make_toolbar:  # just stack canvas and the additional widgets, if any
+            self.output = (frame.canvas, *add_widgets)
+
+        if make_toolbar:  # make toolbar and stack canvas, toolbar, add_widgets
+            self.toolbar = IpywidgetToolBar(frame)
```

Comment:
what if ImageWidget? where is logic for if plot type is imagewidget?

Reply:
`ImageWidget` just uses the `Gridplot` context:

```python
    def show(self, toolbar: bool = True, sidecar: bool = False, sidecar_kwargs: dict = None):
        """
        Show the widget

        Returns
        -------
        OutputContext
        """
        if self.gridplot.canvas.__class__.__name__ == "JupyterWgpuCanvas":
            self._image_widget_toolbar = IpywidgetImageWidgetToolbar(self)

        elif self.gridplot.canvas.__class__.__name__ == "QWgpuCanvas":
            self._image_widget_toolbar = QToolbarImageWidget(self)

        return self.gridplot.show(
            toolbar=toolbar,
            sidecar=sidecar,
            sidecar_kwargs=sidecar_kwargs,
            add_widgets=[self._image_widget_toolbar]
        )
```